packages/office-ui-fabric-react/src/components/TextField/TextField.base.tsx

```diff
@@ -127,69 +105,81 @@ export class TextFieldBase extends React.Component<ITextFieldProps, ITextFieldSt
   /**
    * Gets the current value of the text field.
    */
-  public get value(): string | undefined {
-    return this.state.value;
+  public get value(): string {
+    const value = _getValue(this.props, this.state);
+    if (typeof value === 'number') {
+      return String(value);
+    }
+    return value || '';
   }
 
   public componentDidMount(): void {
-    this._isMounted = true;
     this._adjustInputHeight();
 
     if (this.props.validateOnLoad) {
-      this._validate(this.state.value);
+      this._validate(this.value);
     }
   }
 
-  public componentWillReceiveProps(newProps: ITextFieldProps): void {
-    const { onBeforeChange } = this.props;
-
-    // If old value prop was undefined, then component is controlled and we should
-    //    respect new undefined value and update state accordingly.
-    if (newProps.value !== this.state.value && (newProps.value !== undefined || this.props.value !== undefined)) {
-      if (onBeforeChange) {
-        onBeforeChange(newProps.value);
-      }
-
-      this._id = newProps.id || this._id;
-      this._setValue(newProps.value);
+  public componentWillUnmount() {
+    this._async.dispose();
+  }
 
-      const { validateOnFocusIn, validateOnFocusOut } = newProps;
-      if (!(validateOnFocusIn || validateOnFocusOut)) {
-        this._delayedValidate(newProps.value);
-      }
-    }
+  public getSnapshotBeforeUpdate(prevProps: ITextFieldProps, prevState: ITextFieldState): ITextFieldSnapshot | null {
+    const snapshot: ITextFieldSnapshot = {
+      // TODO: any negative perf impact from always saving this? (involves accessing DOM)
+      selection: [this.selectionStart, this.selectionEnd]
+    };
+    const props = this.props;
 
-    // If component is not currently controlled and defaultValue changes, set value to new defaultValue.
-    if (newProps.defaultValue !== this.props.defaultValue && newProps.value === undefined) {
-      this._setValue(newProps.defaultValue);
+    // TODO: right place for this? would doing it in componentDidUpdate work timing-wise instead?
+    const prevValue = _getValue(prevProps, prevState);
+    const value = this.value;
+    if (prevValue !== value && props.onBeforeChange) {
```

Comment:
I'm pretty sure saving the selection state (above) is a proper use for getSnapshotBeforeUpdate, but this part is more questionable. I also suspect that changing the onBeforeChange logic may be broke MaskedTextField;